### PR TITLE
[MM-59069] Make sure OTP are actual One Time Password

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
Prevents users from being able to reuse the same code twice in a row, One Time Passwords are meant to be used once as the name suggests.

To do this, we provide the library with the last 3 (3 being the size of our window) timestamps used, the library uses those codes to eliminate codes that have already been used, and updates the timestamps list that we then store in our database.

The library takes care of maintaining the timestamp list up to date.

This PR adds a new column in the `Users` table to store those timestamps.

#### Ticket Link
Jira https://mattermost.atlassian.net/browse/MM-59069

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: server/channels/db/migrations/postgres/000126_add_mfa_used_ts_to_users.up.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE Users ADD COLUMN IF NOT EXISTS MfaUsedTimestamps VARCHAR(35) NOT NULL DEFAULT '';
```
Comment (agnivade) [inline server/channels/db/migrations/postgres/000126_add_mfa_used_ts_to_users.up.sql:1]: It seems like we are actually storing an array of integers, and then converting them into a comma separated string and storing a text. Why not have `integer[]` as a data type?

Comment (JulienTant) [reply server/channels/db/migrations/postgres/000126_add_mfa_used_ts_to_users.up.sql:1]: That would be `integer[]` for postgres and `JSON` for MySQL?

Comment (agnivade) [reply server/channels/db/migrations/postgres/000126_add_mfa_used_ts_to_users.up.sql:1]: Indeed, in that case I'd suggest to go with jsonb for postgres, json for MySql. We already have an example for that: see the participants column in Threads table. `(s *SqlThreadStore) MaintainMembership` method contains code to add a participant.

The thing with database schema changes are that once it's done, it's nearly impossible to change them later. So it's a good practice to spend some time to get it right.

File: server/public/model/utils.go [hunk 2]
```diff
@@ -41,8 +41,11 @@ const (
 
 var ErrMaxPropSizeExceeded = fmt.Errorf("max prop size of %d exceeded", maxPropSizeBytes)
 
+//msgp:ignore StringInterface StringSet
```
Comment (agnivade) [inline server/public/model/utils.go:44]: You have to ignore `AppError` as well

**Review by agnivade — APPROVED** [review]
Small change needed. Good to go after that. Approving for speed.

File: server/public/model/utils.go (hunk 2 above)
Comment (JulienTant) [reply server/public/model/utils.go:44]: Thanks!

**Review by esarafianou — APPROVED** [review]
Looks good, thank you!

**Review by lindy65 — APPROVED** [review]
Thanks @JulienTant - smoke tested MFA and looks good

File: server/platform/shared/mfa/mfa.go [hunk 3]
```diff
@@ -120,18 +123,32 @@ func (m *MFA) Deactivate(userId string) error {
...
 	trimmedToken := strings.TrimSpace(token)
 	ok, err := otpConfig.Authenticate(trimmedToken)
 	if err != nil {
 		return false, errors.Wrap(err, "unable to parse the token")
 	}
+	if !ok {
+		return false, nil
+	}
+
+	err = m.store.StoreMfaUsedTimestamps(user.Id, otpConfig.DisallowReuse)
```
Comment (mgdelacroix) [inline server/platform/shared/mfa/mfa.go:148]: Quick question, after calling `.Authenticate`, the `DisallowReuse` property contains only those timestamps inside of the window and the rest are cleared, right?

Comment (mgdelacroix) [reply server/platform/shared/mfa/mfa.go:148]: Discussed offline, the `DisallowReuse` property is updated as part of the call

**Review by mgdelacroix — APPROVED** [review]
LGTM! :+1:

**Comment by amyblais** [comment]
@JulienTant Can you help with a manual cherry-pick PR for ``release-10.0`` branch?

**Pull request merged**